src/vm/circuit.rs

```
@@ -0,0 +1,333 @@
+            let rhs = outputs.get(&(add_gate.rhs as u64));
+
+            match (lhs, rhs) {
+                (Some(l), Some(r)) => {
+                    outputs.insert(add_gate.output as u64, (l + r) % &modulus);
+                }
+                (None, Some(r)) => {
+                    let res = match outputs.get(&(add_gate.output as u64)) {
+                        Some(res) => res,
+                        None => break,
+                    };
+                    // if it is a sub_gate the output index is store in lhs
+                    outputs.insert(add_gate.lhs as u64, (res + &modulus - r) % &modulus);
+                }
+                // there aren't enough gates computed for add_gate to compute
+                // the next gate so we need to compute a mul_gate
+                _ => break,
+            };
+
+            add_gates.next();
+        }
+
+        match mul_gates.next() {
+            Some(mul_gate) => {
+                let lhs = outputs.get(&(mul_gate.lhs as u64));
+                let rhs = outputs.get(&(mul_gate.rhs as u64));
+
+                match (lhs, rhs) {
+                    (Some(l), Some(r)) => {
+                        outputs.insert(mul_gate.output as u64, (l * r) % &modulus);
+                    }
+                    (None, Some(r)) => {
+                        let res = match r.modinv(&modulus) {
+                            Some(inv) => inv,
+                            None => {
+                                // attempted to get the inverse of 0,
+                                break false;
+                            }
+                        };
+                        // if it is a inv_gate the output index is store in lhs
+                        outputs.insert(mul_gate.lhs as u64, res);
+                    }
+                    // a mul_gate can always be computed because it is only computed
+                    // if an add_gate can't
+                    _ => continue,
```

we should panic or return here as an invalid circuit would end up in an infinite loop